composer/trainer/mosaic_fsdp.py

```diff
@@ -0,0 +1,312 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+# Released under BSD 3-Clause License,
+# Copyright (c) Facebook, Inc. and its affiliates.
+
+"""Updates FSDPs _auto_wrap to enable module_kwargs and custom process_group cache."""
+
+import functools
+import warnings
+from typing import Any, Callable, Dict, Set, Tuple, Union, cast
+
+import torch
+import torch.nn as nn
+from torch import distributed
+from torch.distributed import ProcessGroup
+from torch.distributed.fsdp import (BackwardPrefetch, CPUOffload, FullyShardedDataParallel, MixedPrecision,
+                                    ShardingStrategy)
+from torch.distributed.fsdp._utils import _contains_batchnorm, _override_batchnorm_mixed_precision
+from torch.distributed.fsdp.wrap import _or_policy, _wrap, _wrap_batchnorm_individually
+
+from composer.core import Precision
+from composer.utils import dist
+
+__all__ = [
+    'sharding_map',
+    'backward_prefetch_map',
+    'get_torch_dtype',
+    '_get_mixed_precision',
+    '_get_cpu_offload',
+    '_get_process_group',
+    'MosaicFullyShardedDataParallel',
+]
+
+sharding_map = {
+    'NO_SHARD': ShardingStrategy.NO_SHARD,
+    'SHARD_GRAD_OP': ShardingStrategy.SHARD_GRAD_OP,
+    'FULL_SHARD': ShardingStrategy.FULL_SHARD,
+}
+
+backward_prefetch_map = {
+    'NONE': None,
+    'BACKWARD_PRE': BackwardPrefetch.BACKWARD_PRE,
+    'BACKWARD_POST': BackwardPrefetch.BACKWARD_POST,
+}
+
+
+def get_torch_dtype(dtype: Union[Precision, str]):
+    """Convert common string representations of dtypes to torch dtypes."""
+    dtype = dtype.value if isinstance(dtype, Precision) else dtype
+    if dtype in ['float32', 'torch.float32', 'fp32']:
+        return torch.float32
+    elif dtype in ['float16', 'torch.float16', 'half', 'fp16', 'amp', 'amp_fp16']:
+        return torch.float16
+    elif dtype in ['bfloat16', 'bfloat', 'torch.bfloat16', 'bf16', 'amp_bf16']:
+        return torch.bfloat16
+    elif dtype in ['amp_fp8']:
+        # We use torch.bfloat16 by default for amp_fp8 as there is no
```

Comment:
Do we want to throw a warning here for now? + a TODO when fp8 is supported in PyTorch?

Reply:
This was [copy pasta from dist_strategy.py](https://github.com/mosaicml/composer/pull/2006/files/1e9ebe27db16f25c3528dbd8d4c53c6652f49757#diff-dd51f7e9f15c34f3f9e4895d122b92bf4e8e713cdf582a4af5e5e2eef628c0d1L126), but yes adding a warning sounds like a good idea. will do.